src/snark/constraints.rs

```diff
@@ -0,0 +1,534 @@
+
+        let ns = cs.into();
+        let cs = ns.cs();
+
+        let params = get_params::<F, CF>(&cs);
+
+        let obj = f()?;
+
+        // Step 1: use BooleanInputVar to allocate the values as bits
+        // This is to make sure that we are using as few elements as possible
+        let boolean_allocation =
+            BooleanInputVar::new_input(ns!(cs, "boolean"), || Ok(obj.borrow()))?;
+
+        // Step 2: allocating the nonnative field elements as witnesses
+        let mut field_allocation = Vec::<AllocatedNonNativeFieldVar<F, CF>>::new();
+
+        for elem in obj.borrow().iter() {
+            let mut elem_allocated = AllocatedNonNativeFieldVar::<F, CF>::new_witness(
+                ns!(cs, "allocating element"),
+                || Ok(elem),
+            )?;
+
+            // due to the consistency check below
+            elem_allocated.is_in_the_normal_form = true;
+            elem_allocated.num_of_additions_over_normal_form = CF::zero();
+
+            field_allocation.push(elem_allocated);
+        }
+
+        // Step 3: check consistency
+        for (field_bits, field_elem) in boolean_allocation.val.iter().zip(field_allocation.iter()) {
+            let mut field_bits = field_bits.clone();
+            field_bits.reverse();
+
+            // must use lc to save computation
+            for (j, limb) in field_elem.limbs.iter().enumerate() {
+                let bits_slice = if j == 0 {
+                    field_bits[0..params.bits_per_top_limb].to_vec()
+                } else {
+                    field_bits[params.bits_per_top_limb + (j - 1) * params.bits_per_non_top_limb
+                        ..params.bits_per_top_limb + j * params.bits_per_non_top_limb]
+                        .to_vec()
+                };
+
+                let mut lc = LinearCombination::<CF>::zero();
+                let mut cur = CF::one();
+
+                for bit in bits_slice.iter().rev() {
+                    lc = lc + &bit.lc() * cur;
+                    cur.double_in_place();
+                }
+
+                lc = lc - limb.variable;
+                cs.enforce_constraint(lc!(), lc!(), lc).unwrap();
+            }
+        }
+
+        let mut wrapped_field_allocation = Vec::<NonNativeFieldVar<F, CF>>::new();
+        for field_gadget in field_allocation.iter() {
+            wrapped_field_allocation.push(NonNativeFieldVar::Var(field_gadget.clone()));
+        }
+        Ok(Self {
+            val: wrapped_field_allocation,
+        })
+    }
+}
+
+impl<F, CF> FromFieldElementsGadget<F, CF> for NonNativeFieldInputVar<F, CF>
+where
+    F: PrimeField,
+    CF: PrimeField,
+{
+    fn repack_input(src: &Vec<F>) -> Vec<CF> {
+        BooleanInputVar::repack_input(src)
+    }
+
+    fn from_field_elements(src: &Vec<FpVar<CF>>) -> Result<Self, SynthesisError> {
+        // the same routine as the new_input above.
+
+        let cs = src.cs();
```

Comment:
You probably want to handle the case of constants separately from variables (in the former I guess you don't do any constraint generation?)